### ISSUE TITLE
Model accuracy issue despite training with 1000+ images per class

### ISSUE THREAD

**Issue by ShanthiniMichael** [ultralytics/ultralytics#23478]
I have trained the model using more than 1,000 images per class; however, the model is still not producing accurate or reliable classification results and frequently confuses the objects.

I would appreciate your guidance on the possible reasons for this behavior and suggestions to improve model performance. As requested, I will share the relevant details below, including dataset information, annotations, training configuration, and validation outputs, to help with analysis.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ShanthiniMichael, thank you for your interest in Ultralytics 🚀! This is an automated response to help gather the right details quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (it sounds like it may be), please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).🙏  
To help the team triage your “accuracy/confusion despite 1000+ images per class” report, please include:  
-Exact task and mode used (classification vs detection vs segmentation) and the command/code you ran  
-Your dataset structure + class list and how train/val/test are split  
-Representative sample images per class (especially confusing pairs) + a few label/annotation examples  
-Full training logs and final metrics, plus any validation outputs you referenced  
-Your environment details (OS, Python, PyTorch, GPU, CUDA)

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Are the images unique? Or did you augment the same images?

**Comment by ShanthiniMichael** [comment]
@Y-T-G All images used for training are unique, with no duplicates.

**Comment by Y-T-G** [comment]
@ShanthiniMichael Are they augmented versions of the same images?

The accuracy of the model on the validation set is pretty much 100%. So either it overfit because of lack of variations or your inference images don't resemble your training images.

**Comment by ShanthiniMichael** [comment]
The dataset consists entirely of unique images and does not include augmented versions of the same base images. The images were captured at five different camera distances and across approximately twenty different backgrounds to introduce variation in scale and appearance. Each image contains approximately 20 objects, and the complete dataset includes the same set of around 150 objects repeated across images.

Could you please guide me on the next steps to proceed?

**Comment by glenn-jocher** [comment]
Given the validation accuracy is ~100% while real-world inference is poor, the most likely issue is **leakage/memorization**: if the same ~150 physical objects appear across both train/val splits, the model can learn object-specific cues (and backgrounds) that don’t generalize to new objects; the right next step is to build a **true holdout test set** containing completely unseen physical objects (ideally also new capture conditions) and re-check results following the [Model Testing guide](https://docs.ultralytics.com/guides/model-testing/), and if each image contains ~20 objects but you’re doing `classify`, please confirm you actually want a *single label per image*—otherwise you’ll want to crop each object into its own image or switch to detection so the model focuses on the object (your ~75×75 px target is very small relative to 1024×1024, so crops help a lot); if you paste the exact train/predict commands plus `yolo checks` output (and a few misclassified inference images as files, not screenshots), we can pinpoint whether this is split leakage vs. a preprocessing/domain mismatch.

```bash
yolo checks
```

**Comment by ShanthiniMichael** [comment]
YOLO checks response:

```
Ultralytics 8.4.8 🚀 Python-3.9.23 torch-2.8.0+cu128 CPU (Intel Xeon CPU E5-2696 v4 @ 2.20GHz)
Setup complete ✅ (88 CPUs, 125.2 GB RAM, 39.2/69.9 GB disk)

OS                     Linux-5.14.0-635.el9.x86_64-x86_64-with-glibc2.34
Environment            Linux
Python                 3.9.23
Install                git
Path                   /home/admin/Documents/ultralytics/myenv/lib/python3.9/site-packages/ultralytics
RAM                    125.15 GB
Disk                   39.2/69.9 GB
CPU                    Intel Xeon CPU E5-2696 v4 @ 2.20GHz
CPU count              88
GPU                    None
GPU count              None
CUDA                   None

numpy                  ✅ 2.0.2>=1.23.0
matplotlib             ✅ 3.9.4>=3.3.0
opencv-python          ✅ 4.13.0.90>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.13.1>=1.4.1
torch                  ✅ 2.8.0<2.10,>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,<2.10,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.2.2>=5.8.0
polars                 ✅ 1.36.1>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18
```


Train Script:

```
import sys
sys.path.append("ultralytics")
from ultralytics import YOLO

def main():
    model = YOLO("[yolo11s.pt](http://yolo11s.pt/)")  # small model

    model.train(
        data="data.yaml",       # your dataset config
        imgsz=1024,             # high-res for small objects
        epochs=100,             # enough to learn small dataset
        batch=2,                # safe for CPU
        device="cpu",           # '0' for GPU if available
        project="runs",
        name="yolo11_dal_finetune_26_01",
        workers=2,
        scale=1.0,
    )
if __name__ == "__main__":
    main()
```

Detect Script:

```
from ultralytics import YOLO

# Load your trained model
model = YOLO("runs/yolo11_dal_finetune_26_01/weights/[best.pt](http://best.pt/)")  # path to your trained weights

# Run detection
results = model.predict(
    "/home/admin/Downloads/172.jpg",  # path to your image
    imgsz=1024,        # same as training
    conf=0.25,         # confidence threshold
    iou=0.45,          # NMS IoU threshold
    save=True,         # saves output image with boxes
    save_txt=True,      # saves predicted box coordinates
    project="runs/yolo11_test",  # custom folder
    name="my_predictions" 
)

print("Detection complete. Check the 'runs/detect/predict' folder for results.")
```

Data.yaml

```
path: final_dataset_full
train: train
val: valid

nc: 2
names: ['Toor Dal', 'Bengal Gram']
```

[Object 1 - sample 1](https://github.com/user-attachments/assets/f3ad85cf-36e5-4e5f-8a53-1e94338b6d79)

[Object 1 - sample 2](https://github.com/user-attachments/assets/dd2fcc2f-120c-4e7e-be6b-cabf34515dd4)

[Object 2 - sample 1](https://github.com/user-attachments/assets/801d0633-09c3-4559-aa33-73008172d218)

[Object 2 - sample 2](https://github.com/user-attachments/assets/fae8da1a-4c4d-44bb-b7ad-08a765381e7b)

[Output 1](https://github.com/user-attachments/assets/8569825c-a2f1-473d-9705-0fdded2c81ab)

[Output 2](https://github.com/user-attachments/assets/e0c01676-b2a2-4a97-b9f7-e8910a002735)

**Comment by Y-T-G** [comment]
It's overfitting. Your training images are very different from inference images. Simply taking images of the same object from different angles isn't enough to create variations. It has to be real variations using real images from the wild where you're planning to run inference.

**Comment by ShanthiniMichael** [comment]
Okay, I wanted to confirm whether the dataset is causing the issue. Should I provide a dataset similar to the one used for inference for training purposes as well?

**Comment by Y-T-G** [comment]
> Should I provide a dataset similar to the one used for inference for training purposes as well?

Yes

**Comment by ShanthiniMichael** [comment]
Okay, Thank you.

**Comment by ShanthiniMichael** [comment]
I wanted to confirm that for training, I used 1024×1024 px images, is this suitable for learning small objects?

For inference or processing, if cropping is required for small objects, would it be acceptable to use 640×640 px images instead of 1024×1024 px?

**Comment by Y-T-G** [comment]
You should infer at the same image size as training. If you're cropping, then cropped object should be the same size as training images. The goal of cropping should be to make the objects similar in size to training dataset.

**Comment by ShanthiniMichael** [comment]
Okay

**Comment by glenn-jocher** [comment]
Got it—one quick cleanup: in your scripts the model/weight paths should be plain filenames (the Markdown like `[yolo11s.pt](...)` / `[best.pt](...)` will break), and you should remove `sys.path.append("ultralytics")` to avoid accidentally shadowing your installed package; after that, keep `imgsz` the same for `train` and `predict` unless you retrain using the same crop/tiling pipeline you’ll use at inference time (the [detection task docs](https://docs.ultralytics.com/tasks/detect/) cover the expected workflow).

```python
from ultralytics import YOLO

model = YOLO("yolo11s.pt")  # or yolo26s.pt for a newer baseline
# ...
model = YOLO("runs/yolo11_dal_finetune_26_01/weights/best.pt")
```

**Comment by ShanthiniMichael** [comment]
@glenn-jocher Sorry for the confusion. I have used it correctly, but when copying the code, it appears that way.

**Comment by glenn-jocher** [comment]
No worries—if the actual code uses plain strings for `YOLO("yolo11s.pt")` and `YOLO(".../best.pt")`, then the remaining gap is almost certainly **data/domain mismatch + leakage**: since your ~150 physical objects repeat, make sure your splits are **grouped by physical object** (i.e., the same object must not appear in both `train` and `val`/`test`), then evaluate on a true holdout set following the [Model Testing guide](https://docs.ultralytics.com/guides/model-testing/) to confirm real-world performance; also please confirm whether you intend `detect` or `classify` (with ~20 objects/image you’ll usually want detection or per-object crops), as outlined in the [Detection task docs](https://docs.ultralytics.com/tasks/detect/).

**Comment by ShanthiniMichael** [comment]
@glenn-jocher Thank you very much for your guidance. I am now successfully distinguishing between two similar objects.

Next, I would like to handle cases where objects overlap and are partially occluded. Could you please suggest best practices to improve model performance in such scenarios?

Additionally, regarding annotation: when objects overlap or are partially occluded, what is the recommended strategy? Should we annotate every object as long as it is identifiable, even if only a small portion is visible?

**Comment by alobroke** [comment]
Based on the details provided, the main issue appears to be the small relative object size compared to the full image resolution.

Your images are 1024×1024, but objects are only ~75×75 pixels (~0.5% of total area). This makes feature extraction difficult, especially in deeper layers where spatial resolution is reduced.

I recommend the following improvements:

1. Increase input image size during training
Example:
yolo train model=yolo11m-cls.pt imgsz=1280

2. Try larger models
Use yolo11m or yolo11l instead of small models.

3. Crop images around object regions
This increases object prominence and improves feature learning.

4. Enable stronger augmentation
Example:
augment=True
degrees=10
scale=0.5
mosaic=1.0

5. Train longer
Increase epochs to 200–300

6. Verify class balance
Ensure equal samples per class.

Small object classification benefits significantly from higher input resolution and larger models.
It would also help to see confusion matrix and validation accuracy trends.

**Comment by ShanthiniMichael** [comment]
@alobroke Thank you for your response. I am able to achieve around 98% accuracy, even with real-world data.

However, I have a question regarding cases where more than 100 objects appear in a single image and are heavily overlapped with each other.

In such scenarios, how should the annotation be done?
Should I ignore the objects that are not clearly visible, or should I still annotate them if only a small portion is visible?

I will attach a sample image for reference.

![Image](https://github.com/user-attachments/assets/e2c76419-3eaa-43bc-9827-0bb11c50b9fe)

**Comment by alobroke** [comment]
@ShanthiniMichael  Glad to hear that,
To address your question, it’s important to understand how YOLO works internally.
Regarding annotation in heavily overlapping scenarios:

• YOLO learns from visible pixels, so any object that is partially visible and identifiable should be annotated.
• Even if only a small portion is visible, annotating it helps the model learn robust detection under occlusion.
• Do not annotate objects that are completely occluded or cannot be reliably identified.
• Bounding boxes should cover only the visible portion, not the estimated full object behind overlap.
• This approach improves performance in dense and real-world crowded environments and is consistent with datasets like COCO.
• For better results in such scenarios, consider training with higher resolution:

For example, proper training configuration for dense and overlapping objects would include higher resolution and appropriate augmentation:

yolo train \
    model=yolo11m.pt \
    data=data.yaml \
    imgsz=1280 \
    epochs=200 \
    mosaic=1.0 \
    overlap_mask=True

**Comment by ShanthiniMichael** [comment]
Okay, Thank you @alobroke
